src/pushsource/_impl/model/vms.py

```diff
@@ -0,0 +1,57 @@
+import datetime
+
+from .base import PushItem
+from .. import compat_attr as attr
+from .conv import datestr, instance_of_str, instance_of, optional_str, optional
+
+
+@attr.s()
+class VMIRelease(object):
+    """Release metadata associated with a VM image."""
+
+    product = attr.ib(type=str, validator=instance_of_str)
+    """A short product name, for example "RHEL" if this is an image for Red Hat
+    Enterprise Linux."""
+
+    date = attr.ib(type=datetime.date, converter=datestr)
+    """Date at which this image was generated."""
+
+    arch = attr.ib(type=str, validator=instance_of_str)
+    """Architecture of the image, for example "x86_64"."""
+
+    respin = attr.ib(type=int, validator=instance_of(int))
+    """Respin count. 0 for original build of an image, incremented by one for
+    each rebuild."""
+
+    version = attr.ib(type=str, default=None, validator=optional_str)
+    """A <major>.<minor> version string for the image's product version, for example
+    "7.9" if this is an image for Red Hat Enterprise Linux 7.9.x."""
+
+    base_product = attr.ib(type=str, default=None, validator=optional_str)
+    """For layered products, name of the base product for which the image should be used."""
+
+    base_version = attr.ib(type=str, default=None, validator=optional_str)
+    """For layered products, version of the base product for which the image should be used."""
+
+    variant = attr.ib(type=str, default=None, validator=optional_str)
+    """Variant of this image's product (only for products which have variants).  For example,
+    "Server", for Red Hat Enterprise Linux Server."""
+
+    type = attr.ib(type=str, default=None, validator=optional_str)
+    """Release type, typically "ga" or "beta"."""
+
+
+@attr.s()
+class VMIPushItem(PushItem):
+    """A :class:`~pushsource.PushItem` representing a generic VM image.
+
+    This class is used as a parent class for different types of VM images.
+    """
+
+    release = attr.ib(
+        type=VMIRelease, default=None, validator=optional(instance_of(VMIRelease))
```

Comment:
I think the validator here might need to be more complicated.  Currently, it will allow mixing incompatible VMIRelease subclasses and VMIPushItem subclasses.  For example you'll be able to put a generic VMIRelease onto an AmiPushItem, or an AmiRelease onto a VHDPushItem.

I don't think that's a good idea.  Given that, prior to now, AmiPushItem.release has always been an AmiRelease, it would have been fair for all code using it to assume that would be the case, and arguably a backwards-incompatible change if we start allowing objects where this is not the case.  Easier to reason about if the release on a push item is guaranteed to always be of the correct type associated with that item.

Would it work having a class-level private attribute which can be overridden by subclasses?  e.g.

```python
class VMIPushItem(PushItem):
    _RELEASE_TYPE = VMIRelease

    ...
    @release.validator()
    def __validate_release(self, attribute, value):
        ... # validate that value is either None or isinstance(self._RELEASE_TYPE)
```

And then a subclass can ensure that the correct type of release is always used by:

```python
class AmiPushItem(VMIPushItem):
    _RELEASE_TYPE = AmiRelease
```

Reply:
Yes, I like this idea!

I'll change it, thanks!

Reply:
I've fixed it, PTAL again